@xen-orchestra/xapi/network.mjs

```diff
@@ -0,0 +1,76 @@
+import { asyncEach } from '@vates/async-each'
+import { createLogger } from '@xen-orchestra/log'
+
+const { warn } = createLogger('xo:xapi:network')
+
+async function asyncFilter(arr, predicate) {
+  const results = await Promise.all(arr.map(predicate))
+
+  return arr.filter((_v, index) => results[index])
+}
+
+export default class Network {
+  async setMtu(ref, mtu) {
+    const network = await this.getRecord('network', ref)
+    await network.set_MTU(mtu)
+
+    // The MTU will not be updated for VIFs of paused VM, but we can't unplug/replug VIF on a paused VM
+    const pluggedPifs = await asyncFilter(network.PIFs, async pif => {
+      try {
+        return await this.getField('PIF', pif, 'currently_attached')
+      } catch (error) {
+        warn(error)
+        return false
+      }
+    })
+
+    const pluggedVifs = await asyncFilter(network.VIFs, async vif => {
+      try {
+        const attached = await this.getField('VIF', vif, 'currently_attached')
+        if (attached) {
+          const VM = await this.getField('VIF', vif, 'VM')
+          return (await this.getField('VM', VM, 'power_state')) === 'Running'
+        }
+        return false
+      } catch (error) {
+        warn(error)
+        return false
+      }
+    })
+
+    let errorOccurred = false
+
+    for (const { method, networkInterfaces } of [
+      {
+        method: 'VIF.unplug',
+        networkInterfaces: pluggedVifs,
+      },
+      {
+        method: 'PIF.unplug',
+        networkInterfaces: pluggedPifs,
+      }, // unplugging PIFs may be unnecessary
+      {
+        method: 'PIF.plug',
+        networkInterfaces: pluggedPifs,
+      },
+      {
+        method: 'VIF.plug',
+        networkInterfaces: pluggedVifs,
+      },
+    ]) {
+      try {
+        await asyncEach(networkInterfaces, networkInterface => this.callAsync(method, networkInterface), {
+          stopOnError: false,
+        })
+      } catch (error) {
+        // report an error occured, but continue the re-plugging process
+        errorOccurred = true
+        warn(error)
+      }
+    }
+
+    if (errorOccurred) {
+      throw new Error('Some network interfaces could not be updated.')
```

Comment:
I don't think we should throw here, the MTU has been changed.

There is no standard way to report warnings or partial errors in JS, IMHO the warnings in the logs are enough.

@fbeauchamp Any opinions on this?

Reply:
If we don't throw, the user won't have any clues that the settings wasn't applied completly, and checking if the MTU is really 9000 is not obvious. We can't send all the errors since it can be arbitrary long
I think a best effort "something wen wrong, please check the logs" is enough for now